docs/architecture/adr-070-unordered-account.md

```diff
@@ -2,214 +2,134 @@
-  reject the duplicate for you.
-* Finally, worst case if the duplicate transaction is somehow selected for a block
-  proposal, 2nd and all further attempts to evaluate it, will fail during `DeliverTx`,
-  so worst case you just end up filling up block space with a duplicate transaction.
+We facilitate replay protection by storing the unordered sequence, in the Cosmos SDK KV store. Upon transaction ingress, we check if the transaction's unordered
+sequence exists in state, or if the TTL value is stale, i.e. before the current block time. If so, we reject it. Otherwise,
+we add the unordered sequence to state.
 
-```golang
-type TxHash [32]byte
+The state is evaluated during `PreBlocker`. All transactions with an unordered sequence earlier than the current block time
+will be deleted.
 
-const PurgeLoopSleepMS = 500
+```golang
+package unorderedtx
 
-// UnorderedTxManager contains the tx hash dictionary for duplicates checking,
-// and expire them when block production progresses.
-type UnorderedTxManager struct {
-  // blockCh defines a channel to receive newly committed block time
-  blockCh chan time.Time
+import (
+  "fmt"
+  "time"
 
-  mu sync.RWMutex
-	// txHashes defines a map from tx hash -> TTL value, which is used for duplicate
-	// checking and replay protection, as well as purging the map when the TTL is
-	// expired.
-	txHashes map[TxHash]time.Time
-}
+  sdk "github.com/cosmos/cosmos-sdk/types"
 
-func NewUnorderedTxManager() *UnorderedTxManager {
-  m := &UnorderedTxManager{
-		blockCh:  make(chan time.Time, 16),
-		txHashes: make(map[TxHash]time.Time),
-  }
+  "cosmossdk.io/collections"
+  "cosmossdk.io/core/store"
+)
 
- return m
-}
+const (
+  // DefaultMaxTimeoutDuration defines the default maximum duration an unordered transaction
+  // can set.
+  DefaultMaxTimeoutDuration = time.Minute * 10
+)
 
-func (m *UnorderedTxManager) Start() {
-  go m.purgeLoop()
-}
+type UnorderedSequence string
 
-func (m *UnorderedTxManager) Close() error {
-  close(m.blockCh)
-  m.blockCh = nil
-  return nil
+func NewUnorderedSequence(addr string, timestamp uint64) UnorderedSequence {
+  return UnorderedSequence(fmt.Sprintf("%d/%s", timestamp, addr))
 }
 
-func (m *UnorderedTxManager) Contains(hash TxHash)  bool{
-  m.mu.RLock()
-  defer m.mu.RUnlock()
-
-  _, ok := m.txHashes[hash]
-  return ok
+// Manager contains the tx hash dictionary for duplicates checking, and expire
+// them when block production progresses.
+type Manager struct {
+  kvStore store.KVStoreService
 }
 
-func (m *UnorderedTxManager) Size() int {
-  m.mu.RLock()
-  defer m.mu.RUnlock()
+func NewManager(kvStore store.KVStoreService) *Manager {
+  m := &Manager{
+    kvStore: kvStore,
+  }
 
-  return len(m.txHashes)
+  return m
 }
 
-func (m *UnorderedTxManager) Add(hash TxHash, expire time.Time) {
-  m.mu.Lock()
-  defer m.mu.Unlock()
-
-  m.txHashes[hash] = expire
+func (m *Manager) Contains(ctx sdk.Context, sender string, timestamp uint64) (bool, error) {
```

Comment:
this is example code but somewhere here we support multi-signers right?

Reply:
i can document this better, but im treating `sender string` as the possibility of just `cosmosv123521` or `cosmosv1235325,cosmosvio324j234,cosmosvasj324lk2j3`